### PR TITLE
A SAD update

### LINKED ISSUE #156: SAD on the fly atomic UHF hangs
Hi,

With Git: Rev {master} 81f120d the following input file:

    molecule chromium_dimer{
    0 1
    symmetry d2h
       Cr   0.0000   0.0000   0.0000
       Cr   0.0000   0.0000   1.6788
    }
    
    set {
      basis cc-pvdz
    
      guess sad
      scf_type pk
      reference rhf
      INTS_TOLERANCE 0.0
      D_CONVERGENCE 1e-12
      E_CONVERGENCE 1e-12
    }
    
    energy('scf')

hangs for more than 1 hour on

      SCF Guess: Superposition of Atomic Densities via on-the-fly atomic UHF.

pyscf and molpro take only seconds (the total number of orbitals is 86).

Best wishes,
Sebastian

### PR DESCRIPTION
**Pull request by dgasmith**
## Description
The overall goal was to revamp the SCF guess inside Psi to give a bit better convergence and reduce the number of times a user needs to intervene in guess selection. To this end we have improved the speed of SAD, and tweaked default guesses away from CORE. The other half of improving SCF convergence is improving the solver which is delegated to a later date.

## Todos
- [x] Changed SAD to use DF (default) or a Direct JK object. Reduced computational time of SAD by 50-2000x (yes, really).
- [x] SAD guess is now default for RHF, all RHF test cases pass.
- [x] GWH is now the default for ROHF/UHF, cc11 fails and was switched back to CORE.
- [x] Added fractional occupations via electron smearing across a given row. Preliminary tests show that for something like the HF molecule this works really well, but less so for Benzene. However, for UHF test cases this seems to help things dramatically. Turned off by default.
- [x] Closes #156. Now takes ~3 seconds instead of several hours.
- [x] Add Weigend J fit as a fail over fitting basis sets. While this is only J fit, this appears sufficient for SAD and contains auxiliary basis for atoms through Xenon.
- [x] Add secondary failover to Direct if DF is too angular momentum heavy for compiled `libint`.

## Questions
- [x] We are not sold on GWH as the default for ROHF and UHF, the general consensus is that it works better than CORE on the average. Feedback is most welcome.
- [x] There may have been stability issues with SAD in the past. Can anyone recall what these were and provide an example?

## Status
- [x]  Ready to go

### CODE REVIEW

**Comment by susilehtola** [comment]
The way I understand SAD as described in DOI 10.1002/jcc.20393 is that one uses the SAD density matrix as input to a spin-restricted Fock build, and then uses those orbitals for whatever method. There's no reason SAD shouldn't work for ROHF or UHF, because orbitals are available.

Another way to get orbitals for ROHF is to diagonalize the SAD matrix, but this is not as good as an approach because the density changes in the purification.

**Comment by dgasmith** [comment]
@susilehtola Right, thats why I was a little surprised that ~10 ROHF/UHF test cases failed with a SAD guess. Im guessing one can fiddle with this and figure out how to do this a bit better than currently implemented, but it is worrying when were going for consistency now. With RHF its pretty well tested and we cannot find a case where it hurts rather than helps.

Looks like were hitting ERI AM limitation in Travis. Everything should pass when we roll over to direct when the AM of the auxiliary exceeds compiled.

**Comment by susilehtola** [comment]
Btw you don't have Cholesky routines as an alternative to RI? Seems like you do have CD for CC, you could just take the B matrices from the CD and use them for the SCF.

**Comment by dgasmith** [comment]
Sure, we have CD for SCF. Just set 'scf_type cd'.

**Comment by susilehtola** [comment]
So why don't you make that the default for atomic calcs?

**Comment by dgasmith** [comment]
Cholesky vectors can take quite some time to construct. Id rather just roll over to Direct, as we only need 4-5 atomic UHF iterations. Also, when we default to the Weigend basis we only need `AM=5` for all atoms through Rn.

**Comment by robparrish** [comment]
I have always found df/cd to be about the same speed. The only reasons df
is preferred for larger molecules is differentiability, plus lower memory
requirement. For atoms, neither of these are problematic.

Further afield there is no gain for solving hf in the larger basis. Just
solve in a dz type basis and project up.

**Comment by dgasmith** [comment]
Aha, Psi's CD algorithm is not parallelized and is probably why I was seeing CD tensor generation taking longer than the iterations in some cases. We could fix that, it would be nice to avoid an auxiliary basis.

Yea, thats something to consider; however, one thing I have noted is that projecting orbitals that result from a partial Cholesky of the SAD density results in some degradation in the accuracy of the SAD guess itself. This is kind of weird, but might come from the interplay of the projection and how Cholesky weights vectors. Right now the SAD guess takes a small fraction of the SCF time (even for dimers) so I didnt see a need to look into this. Ill add this functionality back in so someone can play with it if they want.

**Comment by robparrish** [comment]
Oh I meant project up atom by atom. This is just like doing the sad guess
in the big basis.

**Comment by kannon92** [comment]
I have found a few cases (Cr2) where DF kills CD.  For the most part, I found that CD takes more time (3-4x) compared to conventional integrals.  For the most part, if you are dealing with transition metals, computing the integrals seems to take a large amount of time and CD requires you to randomly compute different integrals, so the code becomes pretty slow.  If anybody has any ideas of how to speed up this process please tell me.  Right now, I am trying to run a system with 340 basis functions and the cholesky procedure has been running for 70 hours already.  There exists very few DF basis sets with large enough AM for Chromium so I am forced to use Cholesky.   

I have a branch in my fork where I am trying to figure out the major bottlenecks of cholesky and speed it up.  The simplest idea, which is already done, is to compute the cholesky integrals only once and read them from disk.  I basically allow cholesky to read the df_ints_io keyword.  Previously, the CDJK and lib3index do not interact with this keyword, so it would compute the cholesky integrals if choleskify was ever called.  

I agree with @dgasmith to use conventional integrals for the SAD guess.  Cholesky seems to slower than conventional integrals.  Generation of the cholesky integrals is known to be a slow process as shown by Lindh and Aquilante in many of their papers.

**Comment by kannon92** [comment]
Nice job, @dgasmith.

File: src/lib/lib3index/cholesky.cc [hunk 1]
```diff
@@ -162,11 +163,22 @@ void CholeskyMatrix::compute_row(int row, double* target)
...
 
-CholeskyERI::CholeskyERI(boost::shared_ptr<TwoBodyAOInt> integral, double schwarz,
-    double delta, unsigned long int memory) :
-    integral_(integral), schwarz_(schwarz), Cholesky(delta, memory)
+CholeskyERI::CholeskyERI(boost::shared_ptr<BasisSet> basis, double schwarz,
+    double delta, unsigned long int memory, int num_threads) :
+    basisset_(basis), schwarz_(schwarz), Cholesky(delta, memory)
 {
-    basisset_ = integral_->basis();
+    boost::shared_ptr<IntegralFactory> intfact(new IntegralFactory(basisset_,basisset_,basisset_,basisset_));
```
Comment (dgasmith) [inline src/lib/lib3index/cholesky.cc]: Now if the `basisset_` is not explicitly built as cartesian I (non-deterministically, thread or not threaded) get a segfault that originates from `lib/libmints/twobody.cc:633`. Poking around shows no difference between parallelization here and in other parts of the code. Anyone more familiar with the integral code have ideas? @andysim @jturney

**Comment by susilehtola** [comment]
Stupid question, you do reuse integrals on the same shell, don't you? As in J. Chem. Phys. 2003, 118, 9481.

The other issue is that for Cr2 the integrals are horribly slow probably because the integrals engine isn't taking advantage of the general contraction. E.g. cc-pVDZ is [20s16p8d2f|6s5p3d1f] for Cr so you'll end up doing a whole lot of extra work to get those integrals.

**Comment by kannon92** [comment]
@susilehtola, @robparrish was the original author of the code, but I believe that the integrals are recomputed whenever needed.  Rob can correct me if I am wrong.  The algorithm is extremely simple and could use much more improvements.

**Comment by robparrish** [comment]
Yeah this code recomputes based on the function pivot pair. I am aware of
the many advantages of resolving the full shell pair at once, but just
needed a quick reference code to check compressibility when I wrote this.
It could easily be fixed but I have to this point only ever needed df for
production computations. I'm sojourning in lugano next week and can check a
new algorithm in when I've a free hour then.

Cheers
Rob

**Comment by dgasmith** [comment]
A key thing about SAD is that we can abuse it pretty heavily since we are only interested in the density. So lets just do the easy solution: I trimmed `AM=5` out of the Weigend basis to make a `SAD-FIT` basis. While the energy can deviate up to 1% (!) due to the errors in the fitting basis the densities look similar.

For example, hydobromic acid/aug-cc-pV5Z:

```
SAD DF-basis aug-cc-pV5Z-jkfit
   @DF-RHF iter   0: -2573.11546127243219   -2.57312e+03   1.30325e-02
   @DF-RHF iter   6: -2573.05212026634626   -9.92310e-06   3.01845e-06 DIIS
```

```
SAD DF-basis SAD-FIT
   @DF-RHF iter   0: -2572.77288633232229   -2.57277e+03   2.28224e-02
   @DF-RHF iter   6: -2573.05212044674045   -2.02440e-07   7.83243e-07 DIIS
```

This actually represents  a worst-case example as the `SAD-FIT` basis is 10x smaller than `aug-cc-pV5Z-jkfit` for bromine. I think this represents a happy medium, 150 orbital basis functions on a single atom runs in about 3 seconds and we will not have AM issues (Psi4 minimal AM is 4). Going to run with this for now.

Unless someone has ideas of what went wrong with the CDJK spherical vs cartesian issue I'm just going to rip this out.

**Pull request merged**